File: puppyscript/puppy/expr_optim.py
import re

from .ids import is_debug_info

class ExprOptim:
    content: str
    debug_mode: bool

    def __init__(self, file: str, debug_mode: bool):
        with open(file, "r", encoding="utf-8") as f:
            self.content = f.read()
        self.debug_mode = debug_mode

    def remove_too_many_info(self):
        if self.debug_mode:
            result = list()
            prev = False
            for line in self.content.split("\n"):
                if is_debug_info(line):
                    if not prev:
                        result.append(line)
                    prev = True
                else:
                    result.append(line)
                    prev = False
            self.content = "\n".join(result)

    def work(self, output_name: str) -> str:
        self.remove_too_many_info()
        output = output_name + ".expr_opt"
        with open(output, "w", encoding="utf-8") as f:
            f.write(self.content)
        return output

File: puppyscript/cpp-puppy/expr_optim.cpp
//
// Created by impodog on 7/31/2023.
//
#include "processor.h"

std::regex RE_ARRAY_INDEXES("\\b(" WORD ")\\[");

class ExprOptim {
    std::string content;

    void scan_array_indexes(std::string &s) {
        std::smatch results;
        while (std::regex_search(s, results, RE_ARRAY_INDEXES)) {
            auto arr = results[1].str();
            size_t start = results.position(1) + arr.size();
            size_t depth = 0, pos = start;
            do {
                switch (s[pos]) {
                    case '[':
                        depth++;
                        break;
                    case ']':
                        depth--;
                        break;
                    default:
                        break;
                }
                pos++;
            } while (depth);
            auto subscript = s.substr(start + 1, pos - start - 2);
            scan_array_indexes(subscript);

            std::string replacement = "(";
            replacement.append(arr).append(".at_ref ")
                    .append(subscript).append(")").append(".data");
            s.replace(results.position(0), pos - results.position(0), replacement);
        }
    }

    void scan_all_array_indexes() {
        auto lines = split(content, '\n');
        for (auto &line: lines) {
            scan_array_indexes(line);
        }
        content = join(lines, '\n');
    }

public:
    ExprOptim(const path &file) {
        content = read(file);
    }

    std::string work(const std::string &output_name) {
        scan_all_array_indexes();

        std::string output = output_name + ".expr_optim.puppy";
        write(output, content);
        return output;
    }
};

PROCESSOR(expr_optim) {
    ExprOptim expr_optim(file);
    return expr_optim.work(output_name);
}